repeater.ps1

```diff
@@ -1,48 +1,72 @@
-Function Log($msg) {
+function Log($msg) {
 	$date = Get-Date -Format "yyyy/MM/dd HH:mm:ss"
 	$host.ui.WriteErrorLine("[W] $date $msg")
 }
 
-Function RelayMessage($from, $to, $buf, $arrow) {
+Function ReadMessage($stream) {
+	$buf = New-Object byte[] 4
 	$offset = 0
 	while ($offset -lt 4) {
-		$n = $from.Read($buf, $offset, 4 - $offset);
-		if ($n -eq 0) { exit }
+		$n = $stream.Read($buf, $offset, 4 - $offset);
+		if ($n -eq 0) {
+			break
+		}
 		$offset += $n;
 	}
-	$len = (($buf[0] * 256 + $buf[1]) * 256 + $buf[2]) * 256 + $buf[3] + 4
-	Log "[L] $arrow [W] $arrow ssh-agent.exe ($len B)"
-	$len
-	while ($offset -lt $len) {
-		$n = $from.Read($buf, $offset, [Math]::Min($len, $buf.Length) - $offset)
-		if ($n -eq 0) { exit }
-		$offset += $n
-		$to.Write($buf, 0, $offset)
-		$len -= $offset
-		$offset = 0
+	if ($offset -eq 4) {
+		$len = (($buf[0] * 256 + $buf[1]) * 256 + $buf[2]) * 256 + $buf[3] + 4
+		[Array]::Resize([ref]$buf, $len)
+		while ($offset -lt $buf.Length) {
+			$n = $stream.Read($buf, $offset, $buf.Length - $offset)
+			if ($n -eq 0) {
+				break
+			}
+			$offset += $n
+		}
 	}
+	[Array]::Resize([ref]$buf, $offset)	
+	return $buf
 }
 
 Function MainLoop {
 	Try {
-		$buf = New-Object byte[] 8192
+		$ignoreOpenSSHExtensions = $false
+		Try {
+			$sshAgentVersion = (Get-Command -CommandType Application ssh-agent.exe -ErrorAction Stop)[0].Version
+			$ignoreOpenSSHExtensions = ($sshAgentVersion.Major -le 8 -and $sshAgentVersion.Minor -lt 9)
+			Log "ssh-agent.exe version: $($sshAgentVersion.ToString()) (ignoreOpenSSHExtensions: $ignoreOpenSSHExtensions)"
+		}
+		Catch {
+			$ignoreOpenSSHExtensions = $true
+		}
+
 		$ssh_client_in = [console]::OpenStandardInput()
 		$ssh_client_out = [console]::OpenStandardOutput()
 
 		$ver = $PSVersionTable["PSVersion"]
 		Log "ready: PSVersion $ver"
-
-		$buf[0] = 0xff
-		$ssh_client_out.Write($buf, 0, 1)
+		$ssh_client_out.WriteByte(0xff)
 
 		while ($true) {
 			Try {
 				$null = $ssh_client_in.Read((New-Object byte[] 1), 0, 0)
 				$ssh_agent = New-Object System.IO.Pipes.NamedPipeClientStream ".", "openssh-ssh-agent", InOut
 				$ssh_agent.Connect()
 				Log "[W] named pipe: connected"
-				$len = RelayMessage $ssh_client_in $ssh_agent $buf "->"
-				$len = RelayMessage $ssh_agent $ssh_client_out $buf "<-"
+				$buf = ReadMessage $ssh_client_in
+				if ($ignoreOpenSSHExtensions -and $buf.Length -gt 4 -and $buf[4] -eq 0x1b) {
+					$buf = [byte[]](0, 0, 0, 1, 6)
+					Log "ssh <- [L] (5 B) <dummy for OpenSSH ext.>"

```

Comment:
Because it is not `ssh <- [L]`, this message should be `[W] return dummy for OpenSSH ext.` or something .